### ISSUE TITLE
Warning when import model and result in incorrect behavior

### ISSUE THREAD

**Issue by Pannic17** [Unity-Technologies/barracuda-release#225]
I trained a MobileNetV2 model for classification use in TensorFlow with 7 outputs by self-implementation structure. I transformed saved_model via tf2onnx with opset 9 and 11 but I got warning when importing to unity which says "ValidateIntermediateNCHWModelLayers::Squeeze unsupported multi axis Squeeze". 

![21UL2 - classification - Android - Unity 2021 1 9f1 Personal _DX11_ 2021_9_16 20_20_54](https://user-images.githubusercontent.com/56242138/133611125-2cbab06e-4576-46c6-a186-9f25bf75ab3a.png)
![21UL2 - classification - Android - Unity 2021 1 9f1 Personal _DX11_ 2021_9_16 20_20_59](https://user-images.githubusercontent.com/56242138/133611170-9ba35e63-135d-4db1-b2d5-256b19d83554.png)

The model then result in a incorrect behavior with returning almost the same prediction value each time.
It seems like something went wrong with GlobalAveragePooling2D, is there any ways to fix this.

[models.zip](https://github.com/Unity-Technologies/barracuda-release/files/7177574/models.zip)

**Comment by mantasp** [comment]
what operation are you using after `GlobalAveragePool` in your original model? I would suggest to use `Flatten` instead of `Unsqueeze`

**Comment by AlexRibard** [comment]
Yes that is the problem, we do not support multi axis Squeeze.
The solution is to us a `Flatten` layer instead. That said we could detect those cases ourselves and fix it up internally

**Comment by Pannic17** [comment]
The operation I used is `Dropout`, followed by  `Dense `operation to set up the fc layer, I suggest the `Dense `operation is not currently supported

> what operation are you using after `GlobalAveragePool` in your original model? I would suggest to use `Flatten` instead of `Unsqueeze`

**Comment by Pannic17** [comment]
Just wondering that isn't `GlobalAveragePooling2D` do both pooling and flatten in tensorflow? 
I'm not sure whether `Dropout` or `Dense` is not supported, seems like `Squeeze` is not called in my model

> Yes that is the problem, we do not support multi axis Squeeze.
> The solution is to us a `Flatten` layer instead. That said we could detect those cases ourselves and fix it up internally

**Comment by mantasp** [comment]
`GlobalAveragePool` in ONNX produces tensor of the same rank as input (just with spatial dimensions of 1), I would suggest just to add `Flatten` after or before `Dropout` in your source model and check if ONNX converter still inserts `Unsqueeze`

**Comment by Pannic17** [comment]
![tensorflow_mobilenet – model_v2 py 2021_9_17 0_50_40 (2)](https://user-images.githubusercontent.com/56242138/133653369-fb612f84-959a-4d0f-8bd8-6891e7062966.png)
I tried add `Flatten` before and after `Dropout`, but either would work, I also tried replace `Dropout` and `Dense` by `Reshape` and `Activation` but the Warning remains.

![cmd1](https://user-images.githubusercontent.com/56242138/133654267-77d7d920-e3e3-4272-a101-98294025ec7c.png)

I'm not sure which part went wrong and tf2onnx seems retain `Squeeze` operation

> `GlobalAveragePool` in ONNX produces tensor of the same rank as input (just with spatial dimensions of 1), I would suggest just to add `Flatten` after or before `Dropout` in your source model and check if ONNX converter still inserts `Unsqueeze`

**Comment by mantasp** [comment]
would you mind to share your saved model in tf format (or code that builds model structure) so we could play with it on our end?

**Comment by Pannic17** [comment]
[models_tf.zip](https://github.com/Unity-Technologies/barracuda-release/files/7180340/models_tf.zip)
[ckpt.zip](https://github.com/Unity-Technologies/barracuda-release/files/7180344/ckpt.zip)

Sure, there are all the saved models in different format and the model structure implementation file

> would you mind to share your saved model in tf format (or code that builds model structure) so we could play with it on our end?

**Comment by Pannic17** [comment]
Any update on this issue?

**Comment by Pannic17** [comment]
> would you mind to share your saved model in tf format (or code that builds model structure) so we could play with it on our end?

Any update on this issue?
I'm using tf2.6.0 and tf2onnx1.9.2
Do I have to rollback the packages?

**Comment by FlorentGuinier** [comment]
Hey @Pannic17 just to let you know i'm on it.

**Comment by FlorentGuinier** [comment]
Hey @Pannic17 

So it appeared that by default keras is removing the spatial dimensions (aka changing the tensor rank) :
[keras global_average_pooling2d doc](https://keras.io/api/layers/pooling_layers/global_average_pooling2d/)
_keepdims: A boolean, whether to keep the spatial dimensions or not. If keepdims is False (default), the rank of the tensor is reduced for spatial dimensions. If keepdims is True, the spatial dimensions are retained with length 1_

While ONNX is not :
[onnx GlobalAveragePooling doc](https://github.com/onnx/onnx/blob/master/docs/Operators.md#globalaveragepool)
_The output tensor has the same rank as the input. The first two dimensions of output shape are the same as the input (N x C), while the other dimensions are all 1._

Thus tf2onnx insert a Squeeze after GlobalAveragePooling to match Keras behaviour in term of tensor rank.
A simple workaround is thus to set keepdims to true in Keras model definition.
`x = layers.GlobalAveragePooling2D(**keepdims=True**)(x)  # pool + flatten`

Please let us know if the workaround works for you :) !

**Comment by Pannic17** [comment]
> Hey @Pannic17 
> 
> 
> 
> So it appeared that by default keras is removing the spatial dimensions (aka changing the tensor rank) :
> 
> [keras global_average_pooling2d doc](https://keras.io/api/layers/pooling_layers/global_average_pooling2d/)
> 
> _keepdims: A boolean, whether to keep the spatial dimensions or not. If keepdims is False (default), the rank of the tensor is reduced for spatial dimensions. If keepdims is True, the spatial dimensions are retained with length 1_
> 
> 
> 
> While ONNX is not :
> 
> [onnx GlobalAveragePooling doc](https://github.com/onnx/onnx/blob/master/docs/Operators.md#globalaveragepool)
> 
> _The output tensor has the same rank as the input. The first two dimensions of output shape are the same as the input (N x C), while the other dimensions are all 1._
> 
> 
> 
> Thus tf2onnx insert a Squeeze after GlobalAveragePooling to match Keras behaviour in term of tensor rank.
> 
> A simple workaround is thus to set keepdims to true in Keras model definition.
> 
> `x = layers.GlobalAveragePooling2D(**keepdims=True**)(x)  # pool + flatten`
> 
> 
> 
> Please let us know if the workaround works for you :) !

That works perfectly fine!
Thank you
